File: src/main/scala/dev/alexnader/framity/client/assets/overlay/ColoredLike.scala
package dev.alexnader.framity.client.assets.overlay

import dev.alexnader.framity.util.json.Parsers.IdentifierParser
import dev.alexnader.framity.util.json.{JsonParseContext, JsonParseResult, JsonParser}
import net.minecraft.block.BlockState
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry

import scala.jdk.OptionConverters._

object ColoredLike {
  implicit object Parser extends JsonParser[ColoredLike] {
    override def apply(context: JsonParseContext): JsonParseResult[ColoredLike] = {
      context.parse[Identifier].flatMap(id =>
        Registry.BLOCK.getOrEmpty(id)
          .toScala
          .map(block => Right(block.getDefaultState))
          .getOrElse[JsonParseResult[BlockState]](Left(s"Invalid ID: $id"))
          .map(new ColoredLike(_))
      )
    }
  }
}

case class ColoredLike(colorSource: BlockState)

File: src/main/java/dev/alexnader/framity/client/assets/overlay/ColoredLike.java
package dev.alexnader.framity.client.assets.overlay;

import com.mojang.serialization.Codec;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.block.BlockState;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;

@Environment(EnvType.CLIENT)
public class ColoredLike {
    public static final Codec<ColoredLike> CODEC = Identifier.CODEC
        .xmap(
            id -> Registry.BLOCK.get(id).getDefaultState(),
            state -> Registry.BLOCK.getId(state.getBlock())
        )
        .xmap(
            ColoredLike::new,
            coloredLike -> coloredLike.colorSource
        );

    private final BlockState colorSource;

    public ColoredLike(final BlockState colorSource) {
        this.colorSource = colorSource;
    }

    public BlockState colorSource() {
        return colorSource;
    }
}
